cmd/kaspawallet/daemon/server/split_transaction.go

```diff
@@ -0,0 +1,259 @@
+				TransactionID: *consensushashing.TransactionID(splitTransaction.Tx),
+				Index:         0,
+			},
+			UTXOEntry:      utxo.NewUTXOEntry(output.Value, output.ScriptPublicKey, false, constants.UnacceptedDAAScore),
+			DerivationPath: s.walletAddressPath(changeWalletAddress),
+		}
+		totalValue += output.Value
+		totalValue -= feePerInput
+	}
+
+	if totalValue < sentValue {
+		// sometimes the fees from compound transactions make the total output higher than what's available from selected
+		// utxos, in such cases - find more UTXO(s) and use them.
+		additionalUTXOs, totalValueAdded, err := s.moreUTXOsForMergeTransaction(utxos, sentValue-totalValue)
+		if err != nil {
+			return nil, err
+		}
+		utxos = append(utxos, additionalUTXOs...)
+		totalValue += totalValueAdded
+	}
+
+	payments := []*libkaspawallet.Payment{{
+		Address: toAddress,
+		Amount:  sentValue,
+	}}
+	if totalValue > sentValue {
+		payments = append(payments, &libkaspawallet.Payment{
+			Address: changeAddress,
+			Amount:  totalValue - sentValue,
+		})
+	}
+
+	mergeTransactionBytes, err := libkaspawallet.CreateUnsignedTransaction(s.keysFile.ExtendedPublicKeys,
+		s.keysFile.MinimumSignatures, payments, utxos)
+	if err != nil {
+		return nil, err
+	}
+
+	return serialization.DeserializePartiallySignedTransaction(mergeTransactionBytes)
+}
+
+func (s *server) maybeSplitTransaction(transaction *serialization.PartiallySignedTransaction,
+	changeAddress util.Address) ([]*serialization.PartiallySignedTransaction, error) {
+
+	transactionMass, err := s.estimateMassAfterSignatures(transaction)
+	if err != nil {
+		return nil, err
+	}
+
+	if transactionMass < mempool.MaximumStandardTransactionMass {
+		return []*serialization.PartiallySignedTransaction{transaction}, nil
+	}
+
+	splitCount, inputCountPerSplit := s.splitAndInputPerSplitCounts(transaction, transactionMass)
+
+	splitTransactions := make([]*serialization.PartiallySignedTransaction, splitCount)
+	for i := 0; i < splitCount; i++ {
+		startIndex := i * inputCountPerSplit
+		endIndex := startIndex + inputCountPerSplit
+		var err error
+		splitTransactions[i], err = s.createSplitTransaction(transaction, changeAddress, startIndex, endIndex)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return splitTransactions, nil
+}
+
+func (s *server) splitAndInputPerSplitCounts(transaction *serialization.PartiallySignedTransaction, transactionMass uint64) (
+	splitCount, inputPerSplitCount int) {
+	transactionWithoutInputs := transaction.Tx.Clone()
+	transactionWithoutInputs.Inputs = []*externalapi.DomainTransactionInput{}
+	massWithoutInputs := s.txMassCalculator.CalculateTransactionMass(transactionWithoutInputs)
+
+	massForInputs := transactionMass - massWithoutInputs
+
+	inputCount := len(transaction.Tx.Inputs)
+	massPerInput := massForInputs / uint64(inputCount)
+	if massForInputs%uint64(inputCount) > 0 {
+		massPerInput++
+	}
+
+	inputPerSplitCount = int((mempool.MaximumStandardTransactionMass - massWithoutInputs) / massPerInput)
```

Comment:
Maybe rename to input**s**PerSplitCount?

Reply:
Each split tx has only one output right? So wouldn't it make more sense to reduce the mass of one output, instead of all outputs?

Reply:
Using the outputs of the original transaction as an upper bound is confusing from the reader POV (why use such outputs that are unrelated to the actual split transaction), and the calculation might go wrong: e.g. P2SH is 34 bytes long and P2PK is 33 bytes long, so when you're using a multisig wallet and wanna pay to P2PK, and say your original transaction had one output, the output of the split transaction will use a P2SH address, which requires more mass, so your split tx can go above limit.

So if you don't want to export another function for calculating output mass, can't you just make a stub output that will use some change address and then calculate the output mass in the same way you calculated `massForInputs`?